presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java

```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
+    {
+        if (customSplitInfo != null && customSplitInfo.containsKey(HiveUtil.CUSTOM_FILE_SPLIT_CLASS_KEY)) {
+            InputFormat<?, ?> inputFormat = HiveUtil.getInputFormat(configuration,
+                    storage.getStorageFormat().getInputFormat(), false);
+            return Arrays.stream(inputFormat.getClass().getAnnotations())
+                    .map(Annotation::annotationType)
+                    .map(Class::getSimpleName)
+                    .anyMatch(name -> name.equals("UseRecordReaderFromInputFormat"));
+        }
+
+        return false;
+    }
+
+    /**
+     * Checks whether to use File Splitting logic from configured InputFormat or not. It checks for the presence of an
+     * annotation named UseFileSplitsFromInputFormat on the InputFormat class. In addition, there are some additional
+     * checks for Apache Hudi tables.
+     * @param inputFormat InputFormat configured for the table
+     * @param conf Hadoop Configuration
+     * @param tablePath  Base path for the table
+     * @return Boolean indicating whether to use File Splitting logic from Input Format or not.
+     */
+    static boolean shouldUseFileSplitsFromInputFormat(ConnectorSession session, InputFormat<?, ?> inputFormat,
+                                                      Configuration conf, String tablePath)
+    {
+        boolean hasUseSplitsAnnotation = Arrays.stream(inputFormat.getClass().getAnnotations())
+                .map(Annotation::annotationType)
+                .map(Class::getSimpleName)
+                .anyMatch(name -> name.equals("UseFileSplitsFromInputFormat"));
+        return hasUseSplitsAnnotation && (!isHudiTable(inputFormat) ||
+                shouldUseFileSplitsForHudi(session, inputFormat, conf, tablePath));
+    }
+
+    /**
+     * This function does additional checks for Apache Hudi tables, to check if File Splitting logic should be used from
+     * Hudi's input format. For Hudi, we need to use file splits logic in the following scenarios:
+     * 1. If the table type is Merge on Read, which means the InputFormat should extend from HoodieRealtimeInputFormat
+     * 2. If the table is created through Hudi BOOTSTRAP operation.
+     *
+     * @param inputFormat InputFormat configured for the table
+     * @param conf Hadoop Configuration
+     * @param tablePath Base path for the table
+     * @return Boolean indicating whether to use File Splitting logic from Input Format or not.
+     */
+    static boolean shouldUseFileSplitsForHudi(ConnectorSession session, InputFormat<?, ?> inputFormat, Configuration conf,
+                                              String tablePath)
+    {
+        // Set Hudi metadata based listing configurations to be used through input format
+        conf.set(HoodieMetadataConfig.METADATA_ENABLE_PROP, String.valueOf(isPreferMetadataToListHudiFiles(session)));
+        conf.set(HoodieMetadataConfig.METADATA_VALIDATE_PROP, String.valueOf(isHudiMetadataVerificationEnabled(session)));
```

Comment:
static import HoodieMetadataConfig.METADATA_ENABLE_PROP and HoodieMetadataConfig.METADATA_VALIDATE_PROP

Reply:
Done.